<div style="border:solid black 2px; padding: 20px"><h1 style="color:black; margin-bottom:20px">Reviewer's comment v1</h1>
Halo Rizdha !

Perkenalkan saya disini sebagai Code Reviewers, disini dengan senang hati saya akan mereview tugas Rizdha ya.

Saya akan memberikan beberapa komen seperti dibawah ini, tolong jangan dipindah, dirubah, maupun dihapus ya.

Komen yang saya berikan akan muncul dalam warna hijau, kuning, atau pun merah seperti ini:

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>

<div class="alert alert-warning">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Terdapat beberapa catatan.

</div>

<div class="alert alert-danger">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Perlu beberapa perbaikan

</div>

Dan tentu saja Rizdha dapat menjawab saya dengan menggunakan ini

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
    
Baik kak
</div>

# Menganalisis risiko gagal bayar peminjam

Tugasmu adalah menyiapkan laporan untuk divisi kredit suatu bank. Kamu akan mencari tahu pengaruh status perkawinan seorang nasabah dan jumlah anak yang dimilikinya terhadap probabilitas gagal bayar dalam pelunasan pinjaman. Pihak bank sudah memiliki beberapa data mengenai kelayakan kredit nasabah.

Laporanmu akan dipertimbangkan pada saat membuat **penilaian kredit** untuk calon nasabah. **Penilaian kredit** digunakan untuk mengevaluasi kemampuan calon peminjam untuk melunasi pinjaman mereka.

[Dalam notebook ini, kamu dapat menemukan petunjuk dan instruksi singkat serta arahan untuk berpikir. Jangan abaikan petunjuk tersebut karena mereka dirancang untuk membekalimu dengan struktur pengerjaan proyek dan akan membantumu melakukan analisis secara lebih mendalam. Sebelum menyerahkan proyekmu, pastikan kamu menghapus semua petunjuk dan deskripsi yang diberikan kepadamu. Sebaliknya, buat laporan ini terlihat seolah-olah dikirimkan oleh kamu kepada rekan kerja untuk menunjukkan temuanmu - mereka tidak perlu tahu bahwa kamu mendapatkan bantuan eksternal dari kami! Untuk membantumu, kami menempatkan semua petunjuk yang harus kamu hapus dalam tanda kurung siku.]

[Sebelum kamu masuk ke dalam proses analisis datamu, jelaskan tujuan proyek dan hipotesis yang akan kamu uji.]

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>

## Buka file data dan baca informasi umumnya.

[Mulailah dengan mengimpor library dan memuat datanya. Kamu mungkin akan menyadari bahwa kamu memerlukan library tambahan saat kamu mengerjakan proyek ini dan itu merupakan hal yang normal. Hanya saja, pastikan untuk memperbarui bagian ini jika dibutuhkan.]

In [1]:
# Muat semua library
import pandas as pd
import numpy as np
# Muat datanya
credit = pd.read_csv('/datasets/credit_scoring_eng.csv')
credit

children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  Secondary Education             1   
3             3   -4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose  
0         40620.102    purchase of the house  
1         17932.802             car purchase  
2         23341.752    purchase of the house  
3         42820.568  supplementary education  
4         25378.572        to have a wedding  
...             ...                      ...  
21520     35966.698     housing transactions  
21521     24959.969        purchase of a car  
21522     14347.610                 property  
21523     39054.888        buying my own car  
21524     13127.587             to buy a car  

[21525 rows x 12 columns]

## Soal 1. Eksplorasi data

**Deskripsi data**
- `children` - jumlah anak dalam keluarga
- `days_employed` - pengalaman kerja nasabah dalam hari
- `dob_years` - usia nasabah dalam tahun
- `education` - tingkat pendidikan nasabah
- `education_id` - pengidentifikasi untuk tingkat pendidikan nasabah
- `family_status` - status perkawinan
- `family_status_id` - pengidentifikasi untuk status perkawinan nasabah
- `gender` - jenis kelamin nasabah
- `income_type` - jenis pekerjaan
- `debt` - apakah nasabah pernah melakukan gagal bayar pinjaman
- `total_income` - pendapatan bulanan
- `purpose` - tujuan mendapatkan pinjaman

[Sekarang, saatnya menjelajahi data kita. Kamu perlu melihat berapa banyak kolom dan baris yang dimiliki oleh data, serta mencermati beberapa baris data untuk memeriksa potensi masalah dengan data.]

In [2]:
# Mari kita lihat berapa banyak baris dan kolom yang dimiliki oleh dataset kita
credit.shape

(21525, 12)

In [3]:
# Mari tampilkan N baris pertama
credit.head()


children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

[Jelaskan apa yang kamu lihat dan perhatikan dalam sampel data yang kamu tampilkan. Apakah terdapat masalah yang mungkin memerlukan penyelidikan dan perubahan lebih lanjut?]
- terdapat minus pada kolom 'days_employed'
- terdapat perbedaan huruf besar/kecil pada kolom 'education'

In [4]:
# Dapatkan informasi data
print(credit.info())
print()
print(credit.isna().sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_i

In [5]:
credit.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

[Apakah terdapat nilai yang hilang di semua kolom atau hanya di beberapa kolom? Jelaskan secara singkat apa yang kamu lihat dalam 1-2 kalimat.]
- pada kolom 'days_employed' dan 'total_income' jumlah barisnya kurang, hanya 19351 seharusnya 21525. hal ini mengindikasikan adanya nilai yang hilang pada kedua kolom tersebut

In [6]:
# Mari kita lihat tabel yang telah difilter dengan nilai yang hilang di kolom pertama yang mengandung data yang hilang
#credit['days_employed'].value_counts(dropna=False)
credit.loc[credit['days_employed'].isna()]


children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

[Apakah nilai yang hilang tampak simetris? Apakah kita yakin dengan asumsi ini? Jelaskan pendapatmu secara singkat di bagian ini. Kamu mungkin ingin melakukan penyelidikan lebih lanjut dan menghitung nilai yang hilang di semua baris dengan nilai yang hilang untuk memastikan bahwa sampel yang hilang memiliki ukuran yang sama.]
- nilai yang hilang nampak simetris, karena nilai yang hilang pada kolom 'days_employed' dan 'total_income' berada pada baris yang sama

In [7]:
# Mari kita terapkan beberapa kondisi untuk memfilter data dan melihat jumlah baris dalam tabel yang telah difilter.
credit_null = credit.loc[ (credit['days_employed'].isna()) & (credit['total_income'].isna())]

credit_null.shape

(2174, 12)

**Kesimpulan sementara**

[Apakah jumlah baris dalam tabel yang telah difilter sesuai dengan jumlah nilai yang hilang? Kesimpulan apa yang bisa kita buat dari hal ini?]
- telah sesuai karena 21525 - 19351 = 2174, kemungkinan 2174 calon nasabah tersebut tidak mau mengisi bagian 'days_employed' dan 'total_income' karena kedua hal tersebut saling berkaitan

[Hitung persentase nilai yang hilang jika dibandingkan dengan keseluruhan dataset. Apakah nilai yang hilang merupakan bagian data yang cukup besar? Jika demikian, kamu sebaiknya perlu mengisi nilai yang hilang. Untuk melakukannya, pertama-tama kita harus mempertimbangkan apakah data yang hilang bisa jadi disebabkan oleh karakteristik nasabah tertentu, seperti jenis pekerjaan atau yang lainnya. Kamu harus memutuskan karakteristik mana yang menurut *kamu* mungkin merupakan penyebabnya. Kedua, kita harus memeriksa apakah ada ketergantungan nilai yang hilang pada nilai indikator lain dengan kolom-kolom yang mengidentifikasikan karakteristik tertentu nasabah.]
- Data yang hilang cukup besar yakni mencapai 10% dari keseluruhan dataset, karekteristik yang menjadi suspect pada analisa data yang hilang ini berdasarkan 'income_type' karena sebagian orang menganggap bahwa income adalah privasi

[Jelaskan langkahmu selanjutnya dan bagaimana korelasinya dengan kesimpulan yang kamu buat sejauh ini.]
- Melihat jumlah total calon nasabah berdasarkan 'income_type'
- Langkah selanjutnya melihat hubungan 'income_type' dengan 'days_employed' dan 'total_income'
- Melihat persentasenya berdasarkan 'income_type'

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>

In [8]:
# Persentase nilai yang hilang jika dibandingkan dengan keseluruhan dataset
a = credit_null.shape[0]
b = credit.shape[0]

print(a/b*100,'%')

10.099883855981417 %


In [9]:
print(credit['income_type'].unique())
print()
credit['income_type'].value_counts(dropna=False)

['employee' 'retiree' 'business' 'civil servant' 'unemployed'
 'entrepreneur' 'student' 'paternity / maternity leave']



employee                       11119
business                        5085
retiree                         3856
civil servant                   1459
unemployed                         2
entrepreneur                       2
paternity / maternity leave        1
student                            1
Name: income_type, dtype: int64

In [10]:
# Mari kita periksa nasabah yang tidak memiliki data tentang karakteristik yang teridentifikasi dan kolom dengan nilai yang hilang
print(credit.loc[(credit['days_employed'].isna()), 'income_type' ].unique())
print(credit[credit['days_employed'].isna()]['income_type'].value_counts())
print()
print(credit.loc[(credit['total_income'].isna()), 'income_type' ].unique())
print(credit[credit['total_income'].isna()]['income_type'].value_counts())

['retiree' 'civil servant' 'business' 'employee' 'entrepreneur']
employee         1105
business          508
retiree           413
civil servant     147
entrepreneur        1
Name: income_type, dtype: int64

['retiree' 'civil servant' 'business' 'employee' 'entrepreneur']
employee         1105
business          508
retiree           413
civil servant     147
entrepreneur        1
Name: income_type, dtype: int64


In [11]:
# Persentase data yang hilang pada karakteristik 'income_type'
aa = credit['income_type'].value_counts(dropna=False)
ab = credit[(credit['days_employed'].isna()) & (credit['total_income'].isna())]['income_type'].value_counts()
ac = ab/aa*100
pd.DataFrame({'total' : aa, 'Loss Data' : ab, 'Loss Data%' : ac})

total  Loss Data  Loss Data%
business                      5085      508.0    9.990167
civil servant                 1459      147.0   10.075394
employee                     11119     1105.0    9.937944
entrepreneur                     2        1.0   50.000000
paternity / maternity leave      1        NaN         NaN
retiree                       3856      413.0   10.710581
student                          1        NaN         NaN
unemployed                       2        NaN         NaN

[Deksripsikan temuanmu di sini.]

**Kemungkinan penyebab hilangnya nilai dalam data**

[Ajukan gagasanmu mengenai kemungkinan alasan hilangnya nilai-nilai tersebut. Apakah menurutmu mereka hilang secara acak atau apakah terdapat pola-pola tertentu?]
- Berdasarkan analisa awal, data yang hilang jika dilihat berdasarkan karakteristik 'income_type' tidaklah secara acak, dan terdapat pola dimana suspectnya :
- 'employee' merupakan yang tertinggi, menurut saya disebabkan karena karakteristik 'employee' menganggap bahwa income mereka merupakan sebuah privasi
- 'business' merupakan yang tertinggi kedua, menurut saya alasannya adalah privasi ataupun bisnis mereka ada yang tidak beres
- 'retiree' adalah cluster calon nasabah yang sudah pensiun, sehingga suspectnya mereka sudah tidak bekerja dan tidak ada penghasilan lagi sehingga mengosongkan kolom 'days_employed' dan 'total_income'
- 'civil servant' menurut saya alasannya adalah privasi
[Mari kita mulai memeriksa apakah nilai yang hilang bersifat acak.]

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>

In [12]:
# Periksalah distribusinya
credit['income_type'].value_counts(normalize=True)


employee                       0.516562
business                       0.236237
retiree                        0.179141
civil servant                  0.067782
unemployed                     0.000093
entrepreneur                   0.000093
paternity / maternity leave    0.000046
student                        0.000046
Name: income_type, dtype: float64

In [13]:
# Memeriksa distribusi di seluruh dataset
credit['children'].value_counts(normalize=True)


 0     0.657329
 1     0.223833
 2     0.095470
 3     0.015331
 20    0.003531
-1     0.002184
 4     0.001905
 5     0.000418
Name: children, dtype: float64

In [14]:
credit['dob_years'].value_counts(normalize=True)

35    0.028664
40    0.028293
41    0.028200
34    0.028014
38    0.027782
42    0.027735
33    0.026992
39    0.026620
31    0.026016
36    0.025784
44    0.025412
29    0.025319
30    0.025087
48    0.024994
37    0.024948
50    0.023879
43    0.023833
32    0.023693
49    0.023600
28    0.023368
45    0.023089
27    0.022904
56    0.022625
52    0.022485
47    0.022300
54    0.022253
46    0.022067
58    0.021417
57    0.021370
53    0.021324
51    0.020813
59    0.020627
55    0.020581
26    0.018955
60    0.017515
25    0.016585
61    0.016492
62    0.016353
63    0.012497
64    0.012311
24    0.012265
23    0.011800
65    0.009013
66    0.008502
22    0.008502
67    0.007758
21    0.005157
0     0.004692
68    0.004599
69    0.003949
70    0.003020
71    0.002695
20    0.002369
72    0.001533
19    0.000650
73    0.000372
74    0.000279
75    0.000046
Name: dob_years, dtype: float64

In [15]:
credit['education'].value_counts(normalize=True)

secondary education    0.638792
bachelor's degree      0.219187
SECONDARY EDUCATION    0.035865
Secondary Education    0.033031
some college           0.031034
BACHELOR'S DEGREE      0.012729
Bachelor's Degree      0.012451
primary education      0.011614
Some College           0.002184
SOME COLLEGE           0.001347
PRIMARY EDUCATION      0.000790
Primary Education      0.000697
graduate degree        0.000186
Graduate Degree        0.000046
GRADUATE DEGREE        0.000046
Name: education, dtype: float64

In [16]:
credit['education_id'].value_counts(normalize=True)

1    0.707689
0    0.244367
2    0.034564
3    0.013101
4    0.000279
Name: education_id, dtype: float64

**Note**

- Kode 0 = Bachelor's Degree (Sarjana/S1)

- Kode 1 = Secondary Education (Pendidikan Menengah Tingkat Pertama/SMP)

- Kode 2 = Some College (Diploma/D3)

- Kode 3 = Primary Education (Pendidikan Dasar/SD)

- Kode 4 = Graduate Degree (Pascasarjana/S2)

In [17]:
credit['family_status'].value_counts(normalize=True)

married              0.575145
civil partnership    0.194053
unmarried            0.130685
divorced             0.055517
widow / widower      0.044599
Name: family_status, dtype: float64

In [18]:
credit['family_status_id'].value_counts(normalize=True)

0    0.575145
1    0.194053
4    0.130685
3    0.055517
2    0.044599
Name: family_status_id, dtype: float64

**Note**

- Kode 0 = Married (Menikah)

- Kode 1 = Civil Partnership (Persatuan Sipil, secara harafiah bisa diartikan sebagai hubungan yang sah dan terdaftar secara hukum dengan orang yang tidak ada hubungannya dengan kita dalam ikatan pernikahan ataupun keluarga, dalam konteks ini bisa berjenis kelamin pria maupun wanita, mendaftarkan hubungan ini secara legal akan memberikan orang yang bersangkutan sebuah hak yang sah untuk bertanggung jawab penuh terhadap partnernya meskipun tidak berstatus suami istri ataupun bukan saudara dan keluarga).

- Kode 2 = Widow/Widower (Janda/Duda, dalam konteks ini status mereka bukan disebabkan oleh perceraian yang disengaja atau diniatkan oleh kedua pihak, tetapi karena lain hal seperti salah satu pihak mengalami kematian, dll).

- Kode 3 = Divorced (Bercerai, dalam konteks ini status perceraian terjadi atas persetujuan kedua belah pihak/diniatkan).

- Kode 4 = Unmarried (Belum menikah / lajang).

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>

In [19]:
credit['gender'].value_counts(normalize=True)

F      0.661370
M      0.338583
XNA    0.000046
Name: gender, dtype: float64

In [20]:
credit['debt'].value_counts(normalize=True)

0    0.919117
1    0.080883
Name: debt, dtype: float64

**Note**

- Kode 0 = Tidak memiliki utang

- Kode 1 = Memiliki utang

In [21]:
credit['purpose'].value_counts(normalize=True)

wedding ceremony                            0.037027
having a wedding                            0.036098
to have a wedding                           0.035958
real estate transactions                    0.031405
buy commercial real estate                  0.030848
buying property for renting out             0.030337
housing transactions                        0.030337
transactions with commercial real estate    0.030244
housing                                     0.030058
purchase of the house                       0.030058
purchase of the house for my family         0.029779
construction of own property                0.029501
property                                    0.029454
transactions with my real estate            0.029268
building a real estate                      0.029082
buy real estate                             0.028990
building a property                         0.028804
purchase of my own house                    0.028804
housing renovation                          0.

**Kesimpulan sementara**

[Apakah distribusi dalam dataset yang asli mirip dengan distribusi tabel yang telah difilter? Apa artinya hal tersebut untuk kita?]
- Distribusi dataset yang asli mirip dengan distribusi tabel yang telah difilter, hal tersebut menandakan bahwa metode filter kita telah benar

[Jika menurutmu kita belum dapat membuat kesimpulan apa pun, mari kita kembali menyelidiki dataset lebih lanjut. Mari pikirkan alasan lain yang dapat menyebabkan data hilang dan periksa apakah kita dapat menemukan pola tertentu yang dapat membuat kita berpikir bahwa hilangnya nilai-nilai tersebut tidak terjadi secara acak. Karena ini merupakan tugasmu maka bagian ini adalah opsional.]
- Untuk saat ini data yang hilang masih terlihat hilang secara acak, namun saya masih meyakini bahwa data yang hilang tersebut membentuk suatu pola jika kita sudah mentransformasi data ini

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>

In [22]:
# Periksa penyebab dan pola lain yang dapat mengakibatkan nilai yang hilang



**Kesimpulan sementara**

[Apakah pada akhirnya kita dapat memastikan bahwa nilai yang hilang adalah suatu kebetulan? Periksa hal lain yang menurutmu penting di sini.]

In [23]:
# Periksa pola lainnya - jelaskan pola tersebut

**Kesimpulan**

[Apakah kamu menemukan suatu pola? Bagaimana kamu mendapatkan kesimpulan ini?]

[Jelaskan bagaimana kamu akan mengatasi nilai-nilai yang hilang. Pertimbangkan kategori dengan nilai yang hilang.]

[Rencanakan secara singkat langkahmu selanjutnya untuk mentransformasi data. Kamu mungkin perlu mengatasi berbagai jenis masalah: duplikat, pencatatan yang berbeda, sumber data yang salah, dan nilai yang hilang.]

## Transformasi data

[Mari kita perhatikan setiap kolom untuk melihat masalah apa yang mungkin dimiliki mereka.]

[Mulailah dengan menghapus duplikat dan memperbaiki data tentang informasi pendidikan jika diperlukan.]

In [24]:
# Mari kita lihat semua nilai di kolom pendidikan untuk memeriksa ejaan apa yang perlu diperbaiki
print(credit['education'].unique())
print()
print(credit['education'].value_counts())
print()
print('Total Calon Nasabah :', credit['education'].count())

["bachelor's degree" 'secondary education' 'Secondary Education'
 'SECONDARY EDUCATION' "BACHELOR'S DEGREE" 'some college'
 'primary education' "Bachelor's Degree" 'SOME COLLEGE' 'Some College'
 'PRIMARY EDUCATION' 'Primary Education' 'Graduate Degree'
 'GRADUATE DEGREE' 'graduate degree']

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64

Total Calon Nasabah : 21525


In [25]:
# Perbaiki pencatatan jika diperlukan
credit['education']=credit['education'].str.lower()

In [26]:
# Periksa semua nilai di kolom untuk memastikan bahwa kita telah memperbaikinya dengan tepat
print(credit['education'].unique())
print()
print(credit['education'].value_counts())
print()
print('Total Calon Nasabah :', credit['education'].count())

["bachelor's degree" 'secondary education' 'some college'
 'primary education' 'graduate degree']

secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: education, dtype: int64

Total Calon Nasabah : 21525


<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>

[Periksa data kolom `children`]

In [27]:
# Mari kita lihat distribusi nilai pada kolom `children`
credit['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

[Apakah terdapat hal-hal yang aneh di kolom tersebut? Jika jawabannya iya, seberapa tinggi persentase data yang bermasalah? Bagaimana mereka bisa terjadi? Buat keputusan tentang apa yang akan kamu lakukan dengan data ini dan jelaskan alasannya.]
- Iya, karena terdapat calon nasabah yang jumlah anaknya '-1' yang mana hal ini imposible, serta terdapat calon nasabah yang jumlah anaknya '20' dan ini juga perlu dipertanyakan
- Keputusan saya setelah melihat anomali ini yakni mereplace nilai '-1' dengan '1' dan mereplace nilai '20' dengan '2' karena suspect saya hal ini dikarenakan kesalahan input oleh calon nasabah

In [28]:
# [perbaiki data berdasarkan keputusanmu]
credit['children']=credit['children'].replace([-1,20],[1,2])

In [29]:
# Periksa kembali kolom `children` untuk memastikan bahwa semuanya telah diperbaiki
credit['children'].value_counts()


0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>

[Periksa data dalam kolom `days_employed`. Pertama-tama, pikirkan tentang masalah apa yang mungkin ada pada kolom tersebut, serta pikirkan juga apa yang mungkin ingin kamu periksa dan bagaimana kamu akan melakukannya.]
- masalah yang mungkin terjadi berdasarkan analisa saya : ada nilai yang bernilai minus, dan terdapat nilai yang hilang 'NaN'

In [30]:
# Temukan data yang bermasalah di kolom `days_employed` jika memang terdapat masalah dan hitung persentasenya
print(credit['days_employed'].unique())
print()
print(credit['days_employed'].value_counts(dropna=False))

[-8437.67302776 -4024.80375385 -5623.42261023 ... -2113.3468877
 -3112.4817052  -1984.50758853]

 NaN            2174
-1580.622577       1
-4122.460569       1
-2828.237691       1
-2636.090517       1
                ... 
-7120.517564       1
-2146.884040       1
-881.454684        1
-794.666350        1
-3382.113891       1
Name: days_employed, Length: 19352, dtype: int64


In [31]:
# Jumlah data yang hilang
credit['days_employed'].isna().sum()

2174

In [32]:
# Jumlah data yang bernilai negatif
(credit['days_employed']<0).sum()

15906

In [33]:
# Persentase data yang bermasalah pada karakteristik 'days_employed'
ba = len(credit['days_employed'].value_counts(dropna=False))
bb = len(credit['days_employed'])
bc = ba/bb*100
print('total data :', bb)
print('total data yang bermasalah :', ba)
print('persentase data yang bermasalah : {:.2f}%'.format(bc))

total data : 21525
total data yang bermasalah : 19352
persentase data yang bermasalah : 89.90%


[Jika jumlah data yang bermasalah tinggi, hal tersebut mungkin disebabkan oleh beberapa masalah teknis. Kita mungkin perlu mengusulkan alasan paling jelas mengapa hal tersebut dapat terjadi dan bagaimana seharusnya data yang benar, mengingat kita tidak dapat menghapus baris yang bermasalah ini.]
- Jumlah data yang bermasalah mencapai 89.9% hal ini sangat tinggi, langkah saya selanjutnya untuk menyelesaikan ini dengan membuat bilangan negatif menjadi positif dengan menggunakan metode absolute, melihat apakah data terdistribusi secara normal atau tidak, jika tidak maka akan saya filter, kemudian mengganti nilai yang hilang dengan median

In [34]:
# Atasi nilai yang bermasalah, jika ada
# Mengabsolutkan nilai yang negatif
credit['days_employed'] = credit['days_employed'].apply(abs)
credit['days_employed'].describe()

count     19351.000000
mean      66914.728907
std      139030.880527
min          24.141633
25%         927.009265
50%        2194.220567
75%        5537.882441
max      401755.400475
Name: days_employed, dtype: float64

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>

- Hal ini masih aneh karena jumlah hari maksimal pada data ini adalah 401755 atau 1100 tahun!, hal ini imposible. oleh sebab itu saya akan memfilter berdasarkan maksimal manusia bekerja selama 40 tahun atau 14600 hari 

In [35]:
credit.loc[credit['days_employed'] > 14600,'days_employed'] = 14600
print(credit['days_employed'].describe())
print()
print(credit['days_employed'].value_counts())

count    19351.000000
mean      4532.289223
std       5127.487807
min         24.141633
25%        927.009265
50%       2194.220567
75%       5537.882441
max      14600.000000
Name: days_employed, dtype: float64

14600.000000    3462
142.276217         1
433.577224         1
3077.307217        1
3650.076979        1
                ... 
2152.475526        1
6219.157426        1
5873.586442        1
3355.260838        1
1636.419775        1
Name: days_employed, Length: 15890, dtype: int64


- Setelah melakukan filtering data menjadi lebih masuk akal, jumlah orang yang bekerja selama 40 tahun sebanyak 3462 orang, dan sesuai dengan data calon nasabah yang 'retiree' berjumlah 3856. Terdapat sedikit perbedaan karena suspect saya beberapa orang memutuskan untuk pensiun dini
- #Dengan outlier yang signifikan sudah tidak ada lagi, maka langkah selanjutnya adalah mengganti nilai yang hilang dengan mean

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>

In [36]:
# Mengganti nilai yang hilang dengan median
#mean = credit['days_employed'].mean()
#credit['days_employed'] = credit['days_employed'].fillna(value= mean)

In [37]:
# Mengubah data float menjadi integer
#credit['days_employed'] = credit['days_employed'].astype(int)

In [38]:
# Periksa hasilnya - pastikan bahwa masalahnya telah diperbaiki
print(credit['days_employed'].isna().sum())
print()
print(credit['days_employed'].value_counts())

2174

14600.000000    3462
142.276217         1
433.577224         1
3077.307217        1
3650.076979        1
                ... 
2152.475526        1
6219.157426        1
5873.586442        1
3355.260838        1
1636.419775        1
Name: days_employed, Length: 15890, dtype: int64


[Sekarang, mari kita lihat usia nasabah dan mengecek apakah terdapat masalah di sana. Sekali lagi, pikirkan tentang kemungkinan kejanggalan apa yang bisa kita temui dalam kolom ini, misalnya angka usia yang tidak masuk akal.]

In [39]:
# Periksa `dob_years` untuk nilai yang mencurigakan dan hitung persentasenya
credit['dob_years'].value_counts().sort_index()

0     101
19     14
20     51
21    111
22    183
23    254
24    264
25    357
26    408
27    493
28    503
29    545
30    540
31    560
32    510
33    581
34    603
35    617
36    555
37    537
38    598
39    573
40    609
41    607
42    597
43    513
44    547
45    497
46    475
47    480
48    538
49    508
50    514
51    448
52    484
53    459
54    479
55    443
56    487
57    460
58    461
59    444
60    377
61    355
62    352
63    269
64    265
65    194
66    183
67    167
68     99
69     85
70     65
71     58
72     33
73      8
74      6
75      1
Name: dob_years, dtype: int64

[Putuskan apa yang akan kamu lakukan dengan nilai yang bermasalah dan jelaskan alasannya.]
- Terdapat 101 orang dengan usia 0 tahun, dan ini imposible
- Saya akan mengganti usia 0 tahun dengan mean karena tidak memiliki outlier yang signifikan

In [40]:
# Atasi masalah pada kolom `dob_years`, jika terdapat masalah
mean = credit['dob_years'].mean()
credit['dob_years'].replace(0, mean, inplace=True)
credit['dob_years'] = credit['dob_years'].astype(int)
mean

43.29337979094077

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>

In [41]:
# Periksa hasilnya - pastikan bahwa masalahnya telah diperbaiki
print(credit['dob_years'].value_counts())

35    617
43    614
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
22    183
66    183
67    167
21    111
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64


[Sekarang, saatnya memeriksa kolom `family_status`. Periksalah nilai seperti apa yang dimuat di dalam kolom ini dan masalah apa yang mungkin perlu kamu atasi.]

In [42]:
# Mari kita lihat nilai untuk kolom ini
credit['family_status'].value_counts()


married              12380
civil partnership     4177
unmarried             2813
divorced              1195
widow / widower        960
Name: family_status, dtype: int64

In [43]:
# Atasi nilai yang bermasalah di `family_status`, jika ada



In [44]:
# Periksa hasilnya - pastikan nilainya telah diperbaiki


[Sekarang, saatnya memeriksa kolom `gender`. Periksalah nilai seperti apa yang dimuat di dalam kolom ini dan masalah apa yang mungkin perlu kamu atasi]

In [45]:
# Mari kita liat nilai dalam kolom ini
credit['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

- Terdapat anomali yakni XNA, karena hanya terdapat 1 data saja, maka XNA akan didrop

In [46]:
credit.loc[credit['gender'] == 'XNA']

children  days_employed  dob_years     education  education_id  \
10701         0    2358.600502         24  some college             2   

           family_status  family_status_id gender income_type  debt  \
10701  civil partnership                 1    XNA    business     0   

       total_income          purpose  
10701     32624.825  buy real estate

In [47]:
# Atasi nilai-nilai yang bermasalah, jika ada
credit=credit.drop(credit.loc[credit['gender'] == 'XNA'].index)

In [48]:
# Periksa hasilnya - pastikan bahwa masalahnya telah diperbaiki
credit['gender'].value_counts()


F    14236
M     7288
Name: gender, dtype: int64

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>

[Sekarang, saatnya memeriksa kolom `income_type`. Periksalah nilai seperti apa yang dimuat di dalam kolom ini dan masalah apa yang mungkin perlu kamu atasi]

In [49]:
# Mari kita lihat nilai dalam kolom ini
credit['income_type'].value_counts()

employee                       11119
business                        5084
retiree                         3856
civil servant                   1459
unemployed                         2
entrepreneur                       2
paternity / maternity leave        1
student                            1
Name: income_type, dtype: int64

In [50]:
# Atasi nilai yang bermasalah, jika ada

In [51]:
# Periksa hasilnya - pastikan bahwa masalahnya telah diperbaiki



[Sekarang, saatnya melihat apakah ada duplikat di dalam data kita. Jika kita menemukannya, kamu harus memutuskan apa yang akan kamu lakukan dengan duplikat tersebut dan menjelaskan alasannya.]

In [52]:
# Periksa duplikat
credit.duplicated().sum()

71

In [53]:
# Atasi duplikat, jika ada
credit = credit.drop_duplicates()

In [54]:
# Lakukan pemeriksaan terakhir untuk mengecek apakah kita memiliki duplikat
credit.duplicated().sum()

0

In [55]:
# Periksa ukuran dataset yang sekarang kamu miliki setelah manipulasi pertama yang kamu lakukan
credit.shape

(21453, 12)

[Jelaskan dataset barumu: jelaskan secara singkat apa saja perubahannya dan seberapa besar persentase perubahannya, jika ada.]
- Memperbaiki kolom 'education' dengan menyamakannya menjadi huruf kecil
- Pada kolom 'children' melakukan replace pada nilai abnormal dimana '-1' diganti dengan '1' dan '20' diganti dengan '2'
- Pada kolom 'days_employed' mengubah nilai negatif menjadi nilai absolute, dan memfilter anomali data dimana jika jumlah hari bekerjanya lebih dari 14600/40tahun akan diubah menjadi 14600 hari, hal ini berdasarkan perkiraan jumlah hari maksimal manusia bekerja
- Pada Kolom 'dob_years' terdapat anomali data dimana usia seseorang 0tahun saat mengajukan kredit, saya mengganti nilai '0' tersebut dengan metode mean karena outliernya tidak signifikan
- Pada kolom 'gender' terdapat 1 data anomali yakni 'XNA', karena hanya 1 data saja maka saya mendrop data tersebut dari dataset
- Terdapat 71 data yang terindikasi merupakan duplikat, dan kemudian mendrop duplikat tersebut
- Dataset yang baru memiliki 21453 baris, sedangkan dataset yang lama memiliki 21525 baris. Terjadi penurunan karena dalam proses analisa beberapa data didrop untuk dapat menghasilkan analisa yang lebih baik

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>

# Bekerja dengan nilai yang hilang

[Untuk mempercepat pekerjaan dengan sejumlah data, kamu mungkin ingin menggunakan dictionary untuk beberapa nilai yang memiliki ID. Jelaskan mengapa dan dictionary manakah yang akan kamu gunakan.]
- Berdasarkan dataset yang kita miliki terdapat 2 kolom yang memiliki ID, yakni 'education_id' dan 'family_id'

In [56]:
# Temukan dictionary
def new_dict (data, col_id, col) :
        result = data[[col_id, col]]
        result = result.drop_duplicates().reset_index(drop=True)
        return result
    
education_dict = new_dict(credit, 'education_id', 'education')
family_status_dict = new_dict(credit, 'family_status_id', 'family_status')

education_dict, family_status_dict

(   education_id            education
 0             0    bachelor's degree
 1             1  secondary education
 2             2         some college
 3             3    primary education
 4             4      graduate degree,
    family_status_id      family_status
 0                 0            married
 1                 1  civil partnership
 2                 2    widow / widower
 3                 3           divorced
 4                 4          unmarried)

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>

### Memperbaiki nilai yang hilang di `total_income`

[Jelaskan secara singkat kolom dengan nilai yang hilang manakah yang perlu kamu tangani. Jelaskan bagaimana kamu akan memperbaikinya.]


[Mulai dengan mengatasi total nilai pendapatan yang hilang. Buat kategori usia untuk nasabah. Buat kolom baru yang memuat kategori usia. Strategi ini dapat membantu untuk menghitung total nilai pendapatan.]


In [57]:
# Mari kita tulis sebuah fungsi untuk menghitung kategori usia
def age_group(age):
    if age <=20:
        value='0-20 years old'
    elif age >=21 and age <=30:
        value='21-30 years old'
    elif age >=31 and age <=40:
        value='31-40 years old'
    elif age >=41 and age <=50:
        value='41-50 years old'
    elif age >=51 and age <=60:
        value='51-60 years old'
    else:
        value='> 60 years old'
    return(value)   
    

In [58]:
# Lakukan pengujian untuk melihat apakah fungsimu bekerja atau tidak
print(age_group(24))

21-30 years old


In [59]:
# Buatlah kolom baru berdasarkan fungsi
credit['age_group'] = credit['dob_years'].apply(age_group)


<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>

In [60]:
# Periksa bagaimana nilai di dalam kolom baru
credit


children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  secondary education             1   
3             3    4124.747207         32  secondary education             1   
4             0   14600.000000         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1    4529.316663         43  secondary education             1   
21521         0   14600.000000         67  secondary education             1   
21522         1    2113.346888         38  secondary education             1   
21523         3    3112.481705         38  secondary education             1   
21524         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose        age_group  
0         40620.102    purchase of the house  41-50 years old  
1         17932.802             car purchase  31-40 years old  
2         23341.752    purchase of the house  31-40 years old  
3         42820.568  supplementary education  31-40 years old  
4         25378.572        to have a wedding  51-60 years old  
...             ...                      ...              ...  
21520     35966.698     housing transactions  41-50 years old  
21521     24959.969        purchase of a car   > 60 years old  
21522     14347.610                 property  31-40 years old  
21523     39054.888        buying my own car  31-40 years old  
21524     13127.587             to buy a car  31-40 years old  

[21453 rows x 13 columns]

[Pikirkan tentang faktor-faktor yang biasanya bergantung pada pendapatan. Pada akhirnya, kamu akan mengetahui apakah kamu harus menggunakan nilai rata-rata atau median untuk mengganti nilai yang hilang. Untuk membuat keputusan ini, kamu mungkin ingin melihat distribusi faktor-faktor yang kamu identifikasi sebagai dampak dari pendapatan seseorang.]

[Buatlah tabel yang hanya memuat data tanpa nilai yang hilang. Data ini akan digunakan untuk memperbaiki nilai yang hilang.]

In [61]:
# Buat tabel tanpa nilai yang hilang dan tampilkan beberapa barisnya untuk memastikan semuanya berjalan dengan baik
new_table = credit.loc[ (credit['days_employed'].isna() == False) & (credit['total_income'].isna() == False)]
new_table

children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  secondary education             1   
3             3    4124.747207         32  secondary education             1   
4             0   14600.000000         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1    4529.316663         43  secondary education             1   
21521         0   14600.000000         67  secondary education             1   
21522         1    2113.346888         38  secondary education             1   
21523         3    3112.481705         38  secondary education             1   
21524         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose        age_group  
0         40620.102    purchase of the house  41-50 years old  
1         17932.802             car purchase  31-40 years old  
2         23341.752    purchase of the house  31-40 years old  
3         42820.568  supplementary education  31-40 years old  
4         25378.572        to have a wedding  51-60 years old  
...             ...                      ...              ...  
21520     35966.698     housing transactions  41-50 years old  
21521     24959.969        purchase of a car   > 60 years old  
21522     14347.610                 property  31-40 years old  
21523     39054.888        buying my own car  31-40 years old  
21524     13127.587             to buy a car  31-40 years old  

[19350 rows x 13 columns]

In [62]:
# Perhatikan nilai rata-rata untuk pendapatan berdasarkan faktor yang telah kamu identifikasi
pd.pivot_table(new_table, columns = 'income_type', values = 'total_income', aggfunc= 'mean')

income_type       business  civil servant      employee  entrepreneur  \
total_income  32386.741818   27343.729582  25820.841683     79866.103   

income_type   paternity / maternity leave       retiree   student  unemployed  
total_income                     8612.661  21940.394503  15712.26  21014.3605

In [63]:
# Perhatikan nilai median untuk pendapatan berdasarkan faktor yang telah kamu identifikasi
pd.pivot_table(new_table, columns = 'income_type', values = 'total_income', aggfunc= 'median')

income_type     business  civil servant    employee  entrepreneur  \
total_income  27571.0825     24071.6695  22815.1035     79866.103   

income_type   paternity / maternity leave    retiree   student  unemployed  
total_income                     8612.661  18962.318  15712.26  21014.3605

[Ulangi perbandingan tersebut untuk beberapa faktor. Pastikan kamu mempertimbangkan berbagai aspek dan menjelaskan proses berpikirmu.]



[Buat keputusan tentang karakteristik yang paling menentukan pendapatan dan apakah kamu akan menggunakan median atau rata-rata. Jelaskan mengapa kamu membuat keputusan ini]
- Menurut saya karakteristik 'income_type' yang paling menentukan pendapatan, dan persebarannya sesuai dengan logika bahwa 'enterpreneur' yang memiliki income tertinggi, dan 'paternity / maternity leave' yang memiliki income terendah
- Karena outlier data cukup signifikan, dan hasil median jika dibanding dengan mean cukup mirip. Pada analisa ini saya menggunakan median untuk mengisi nilai yang hilang

In [64]:
#  Tulis fungsi yang akan kita gunakan untuk mengisi nilai yang hilang
def fillna_from_median_group (dataframe, col_missing, col_group):
    median = pd.pivot_table(dataframe, columns=col_group, values=col_missing, aggfunc='median')

    for column in median.columns:
        dataframe.loc[(credit[col_group]==column) & (dataframe[col_missing].isnull()), col_missing] = median[column][0]

    return dataframe

In [65]:
# Memeriksa bagaimana nilai di dalam kolom baru
credit['total_income'].value_counts()

17312.717    2
31791.384    2
42413.096    2
11927.512    1
55362.332    1
            ..
27715.458    1
23834.534    1
26124.613    1
28692.182    1
41428.916    1
Name: total_income, Length: 19347, dtype: int64

In [66]:
# Terapkan fungsi tersebut ke setiap baris
fillna_from_median_group(credit, col_missing='total_income', col_group='income_type')

children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  secondary education             1   
3             3    4124.747207         32  secondary education             1   
4             0   14600.000000         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1    4529.316663         43  secondary education             1   
21521         0   14600.000000         67  secondary education             1   
21522         1    2113.346888         38  secondary education             1   
21523         3    3112.481705         38  secondary education             1   
21524         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose        age_group  
0         40620.102    purchase of the house  41-50 years old  
1         17932.802             car purchase  31-40 years old  
2         23341.752    purchase of the house  31-40 years old  
3         42820.568  supplementary education  31-40 years old  
4         25378.572        to have a wedding  51-60 years old  
...             ...                      ...              ...  
21520     35966.698     housing transactions  41-50 years old  
21521     24959.969        purchase of a car   > 60 years old  
21522     14347.610                 property  31-40 years old  
21523     39054.888        buying my own car  31-40 years old  
21524     13127.587             to buy a car  31-40 years old  

[21453 rows x 13 columns]

In [67]:
# Periksa apakah kita mendapatkan kesalahan
credit['total_income'].isna().sum()

0

In [68]:
credit['total_income'].value_counts()

22815.1035    1070
27571.0825     501
18962.3180     387
24071.6695     145
42413.0960       2
              ... 
45484.1090       1
27715.4580       1
23834.5340       1
26124.6130       1
41428.9160       1
Name: total_income, Length: 19350, dtype: int64

In [69]:
pd.pivot_table(credit, columns = 'income_type', values = 'total_income')

income_type       business  civil servant      employee  entrepreneur  \
total_income  31911.531001     27018.0956  25530.681104     79866.103   

income_type   paternity / maternity leave      retiree   student  unemployed  
total_income                     8612.661  21640.17577  15712.26  21014.3605

In [70]:
credit['total_income'].describe()

count     21453.000000
mean      26450.780609
std       15710.267466
min        3306.762000
25%       17219.352000
50%       22815.103500
75%       31331.009000
max      362496.645000
Name: total_income, dtype: float64

[Jika kamu menemukan kesalahan dalam menyiapkan nilai untuk data yang hilang, kemungkinan ada sesuatu yang istimewa tentang data untuk kategori tersebut. Coba pikirkan - kamu mungkin ingin memperbaiki beberapa hal secara manual, jika terdapat cukup data untuk menemukan median/rata-rata.]


In [71]:
# Ganti nilai yang hilang jika terdapat kesalahan


[Setelah kamu selesai dengan `total_income`, periksa apakah jumlah total nilai di kolom ini sesuai dengan jumlah nilai di kolom lain.]

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>

###  Memperbaiki nilai di `days_employed`

[Pikirkan tentang parameter yang dapat membantumu memperbaiki nilai yang hilang di kolom ini. Pada akhirnya, kamu akan mengetahui apakah kamu harus menggunakan nilai rata-rata atau median untuk mengganti nilai yang hilang. Kamu mungkin perlu melakukan penelitian yang sama dengan yang kamu lakukan saat memperbaiki data di kolom sebelumnya.]
- Menurut saya karakteristik 'age_group' yang paling menentukan jumlah hari bekerja, dan persebarannya sesuai dengan logika bahwa semakin tua seseorang maka semakin banyak jumlah hari bekerjanya

In [72]:
# Distribusi median dari `days_employed` berdasarkan parameter yang kamu identifikasi
pd.pivot_table(new_table, columns = 'age_group', values = 'days_employed', aggfunc= 'median')

age_group      0-20 years old  21-30 years old  31-40 years old  \
days_employed      695.968951      1064.284435      1630.193189   

age_group      41-50 years old  51-60 years old  > 60 years old  
days_employed      2199.733119      6481.039027         14600.0

In [73]:
# Distribusi rata-rata dari `days_employed` berdasarkan parameter yang kamu identifikasi
pd.pivot_table(new_table, columns = 'age_group', values = 'days_employed', aggfunc= 'mean')

age_group      0-20 years old  21-30 years old  31-40 years old  \
days_employed      673.648361      1318.375904      2180.256329   

age_group      41-50 years old  51-60 years old  > 60 years old  
days_employed      3285.548525      7990.781243    12376.958765

[Tentukan apa yang akan kamu gunakan: rata-rata atau median. Jelaskan mengapa.]
- Karena outlier data cukup signifikan, dan hasil median jika dibanding dengan mean cukup mirip. Pada analisa ini saya menggunakan median untuk mengisi nilai yang hilang

In [74]:
# Mari tulis fungsi yang menghitung rata-rata atau median (tergantung keputusanmu) berdasarkan parameter yang kamu identifikasi
def fillna_from_median_group (dataframe, col_missing, col_group):
    median = pd.pivot_table(dataframe, columns=col_group, values=col_missing, aggfunc='median')

    for column in median.columns:
        dataframe.loc[(credit[col_group]==column) & (dataframe[col_missing].isnull()), col_missing] = median[column][0]

    return dataframe

In [75]:
# Periksa apakah fungsimu bekerja
fillna_from_median_group(credit, col_missing='days_employed', col_group='age_group')


children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  secondary education             1   
3             3    4124.747207         32  secondary education             1   
4             0   14600.000000         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1    4529.316663         43  secondary education             1   
21521         0   14600.000000         67  secondary education             1   
21522         1    2113.346888         38  secondary education             1   
21523         3    3112.481705         38  secondary education             1   
21524         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose        age_group  
0         40620.102    purchase of the house  41-50 years old  
1         17932.802             car purchase  31-40 years old  
2         23341.752    purchase of the house  31-40 years old  
3         42820.568  supplementary education  31-40 years old  
4         25378.572        to have a wedding  51-60 years old  
...             ...                      ...              ...  
21520     35966.698     housing transactions  41-50 years old  
21521     24959.969        purchase of a car   > 60 years old  
21522     14347.610                 property  31-40 years old  
21523     39054.888        buying my own car  31-40 years old  
21524     13127.587             to buy a car  31-40 years old  

[21453 rows x 13 columns]

In [76]:
credit['days_employed'].isna().sum()


0

In [77]:
credit['days_employed'].value_counts()

14600.000000    3671
1630.193189      562
2199.733119      517
6481.039027      466
1064.284435      345
                ... 
6219.157426        1
5873.586442        1
3355.260838        1
472.196656         1
1636.419775        1
Name: days_employed, Length: 15891, dtype: int64

In [78]:
pd.pivot_table(credit, columns = 'age_group', values = 'days_employed')

age_group      0-20 years old  21-30 years old  31-40 years old  \
days_employed      675.708723      1294.365617      2126.324807   

age_group      41-50 years old  51-60 years old  > 60 years old  
days_employed      3181.038033      7835.396088    12595.498566

[Setelah kamu selesai dengan `total_income`, periksa apakah jumlah total nilai di kolom ini sesuai dengan jumlah nilai di kolom lain.]

In [79]:
# Periksa entri di semua kolom - pastikan kita memperbaiki semua nilai yang hilang
credit.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
age_group           0
dtype: int64

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>

## Pengkategorian data

[Untuk menjawab pertanyaan dan menguji hipotesis, kamu akan bekerja dengan data yang telah dikategorikan. Lihat pertanyaan-pertanyaan yang diajukan kepadamu dan yang harus kamu jawab. Pikirkan tentang data mana yang perlu dikategorikan untuk menjawab pertanyaan-pertanyaan ini. Di bawah ini, kamu akan menemukan templat yang bisa kamu gunakan untuk mengkategorikan data. Proses pertama mencakup data teks; yang kedua membahas data numerik yang perlu dikategorikan. Kamu dapat menggunakan kedua petunjuk yang disarankan atau tidak sama sekali - semuanya terserahmu.]

[Terlepas dari keputusanmu untuk mengatasi pengkategorian, pastikan bahwa kamu secara lugas memberikan penjelasan tentang mengapa kamu membuat keputusan tersebut. Ingat: ini adalah pekerjaanmu dan hanya kamu yang berhak membuat segala keputusan.]


In [80]:
# Tampilkan nilai data yang kamu pilih untuk pengkategorian
credit['purpose'].describe()


count                21453
unique                  38
top       wedding ceremony
freq                   791
Name: purpose, dtype: object

[Mari kita periksa nilai unik]

In [81]:
# Periksa nilai unik
credit['purpose'].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

[Kelompok utama apakah yang dapat kamu identifikasi berdasarkan nilai uniknya?]

[Kita akan mengategorikan data kita berdasarkan topik ini.]


In [82]:
# Mari kita tulis sebuah fungsi untuk mengategorikan data berdasarkan topik umum
def purpose_group(row):
    if 'car' in row:
        return 'car'
    elif 'hous' in row or 'prop' in row or 'real est' in row:
        return 'real estate'
    elif 'wedd' in row:
        return 'wedding'
    elif 'educ' in row or 'uni' in row:
        return 'education'
    else:
        return 'Unidentified'

In [83]:
# Buat kolom yang memuat kategori dan hitung nilainya
credit['purpose_group']=credit['purpose'].apply(purpose_group)


[Jika kamu memutuskan untuk mengategorikan data numerik, kamu juga harus membuat kategori untuk data tersebut.]

In [84]:
# Lihat semua data numerik di kolom yang kamu pilih untuk pengkategorian
credit[['purpose','purpose_group']].value_counts()


purpose                                   purpose_group
wedding ceremony                          wedding          791
having a wedding                          wedding          768
to have a wedding                         wedding          765
real estate transactions                  real estate      675
buy commercial real estate                real estate      661
housing transactions                      real estate      652
buying property for renting out           real estate      651
transactions with commercial real estate  real estate      650
purchase of the house                     real estate      646
housing                                   real estate      646
purchase of the house for my family       real estate      638
construction of own property              real estate      635
property                                  real estate      633
transactions with my real estate          real estate      627
building a real estate                    real estate      624

In [85]:
# Dapatkan kesimpulan statistik untuk kolomnya
credit['purpose_group'].value_counts()

real estate    10810
car             4306
education       4013
wedding         2324
Name: purpose_group, dtype: int64

[Tentukan rentang apa yang akan kamu gunakan untuk pengelompokan dan jelaskan alasannya.]

In [86]:
credit['total_income'].describe()

count     21453.000000
mean      26450.780609
std       15710.267466
min        3306.762000
25%       17219.352000
50%       22815.103500
75%       31331.009000
max      362496.645000
Name: total_income, dtype: float64

In [87]:
# Buat fungsi yang melakukan pengkategorian menjadi kelompok numerik yang berbeda berdasarkan rentang
def income_group(income):
    if income <= 10000:
        return 'lower income'
    if income > 10000 and income <= 30000:
        return 'middle income'
    if income > 30000 and income <= 50000:
        return 'middle up income'
    if income > 50000 and income <= 70000:
        return 'upper income'
    if income > 70000:
        return 'rich'


In [88]:
print(income_group(37240))

middle up income


In [89]:
# Buat kolom yang memuat kategori
credit['income_group'] = credit['total_income'].apply(income_group)

In [90]:
# Hitung setiap nilai kategori untuk melihat pendistribusiannya
credit['income_group'].value_counts()

middle income       14608
middle up income     4598
upper income          942
lower income          926
rich                  379
Name: income_group, dtype: int64

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>

## Memeriksa hipotesis


**Apakah terdapat korelasi antara memiliki anak dengan probabilitas melakukan gagal bayar pinjaman?**

In [91]:
# Periksa data anak dan data gagal bayar pinjaman
pd.pivot_table(credit,index='children', columns='debt', values='gender',aggfunc='count')


debt            0       1
children                 
0         13027.0  1063.0
1          4410.0   445.0
2          1926.0   202.0
3           303.0    27.0
4            37.0     4.0
5             9.0     NaN

In [92]:
children_debt=pd.pivot_table(credit,index='children', columns='debt', values='gender',aggfunc='count').fillna(0)
children_debt

debt            0       1
children                 
0         13027.0  1063.0
1          4410.0   445.0
2          1926.0   202.0
3           303.0    27.0
4            37.0     4.0
5             9.0     0.0

In [93]:
# Hitung persentase gagal bayar berdasarkan jumlah anak
children_debt['total']=children_debt[0]+children_debt[1]
children_debt['percent 0']=round(children_debt[0]/children_debt['total']*100,1).astype(str)+'%'
children_debt['percent 1']=round(children_debt[1]/children_debt['total']*100,1).astype(str)+'%'
children_debt

debt            0       1    total percent 0 percent 1
children                                              
0         13027.0  1063.0  14090.0     92.5%      7.5%
1          4410.0   445.0   4855.0     90.8%      9.2%
2          1926.0   202.0   2128.0     90.5%      9.5%
3           303.0    27.0    330.0     91.8%      8.2%
4            37.0     4.0     41.0     90.2%      9.8%
5             9.0     0.0      9.0    100.0%      0.0%

**Kesimpulan**

[Tulis kesimpulanmu berdasarkan manipulasi dan pengamatan yang kamu lakukan.]
- Tidak dapat ditarik kesimpulan berdasarkan korelasi antara jumlah anak dengan hutang, berdasarkan data ini calon nasabah yang memiliki banyak anak, tidak memiliki hutang. Sehingga belum bisa menemukan pola dengan cara ini

**Apakah terdapat korelasi antara status keluarga dengan probabilitas melakukan gagal bayar pinjaman?**

In [94]:
# Periksa data status keluarga dan data gagal bayar pinjaman
family_debt=pd.pivot_table(credit,index='family_status', columns='debt', values='gender',aggfunc='count').fillna(0)
family_debt


debt                   0    1
family_status                
civil partnership   3762  388
divorced            1110   85
married            11408  931
unmarried           2536  274
widow / widower      896   63

In [95]:
# Hitung persentase gagal bayar berdasarkan status keluarga
family_debt['total']=family_debt[0]+family_debt[1]
family_debt['percent 0']=round(family_debt[0]/family_debt['total']*100,1).astype(str)+'%'
family_debt['percent 1']=round(family_debt[1]/family_debt['total']*100,1).astype(str)+'%'
family_debt

debt                   0    1  total percent 0 percent 1
family_status                                           
civil partnership   3762  388   4150     90.7%      9.3%
divorced            1110   85   1195     92.9%      7.1%
married            11408  931  12339     92.5%      7.5%
unmarried           2536  274   2810     90.2%      9.8%
widow / widower      896   63    959     93.4%      6.6%

**Kesimpulan**

[Tulis kesimpulanmu berdasarkan manipulasi dan pengamatanmu.]
- Tidak dapat ditarik kesimpulan berdasarkan korelasi antara status keluarga dengan hutang. Yang bisa kita tarik kesimpulan calon nasabah yang belum menikah memiliki persentase potensi gagal bayang terbesar '9.8%' sedang yang bercerai memiliki persentase potensi gagal bayar terkecil '7.1%'

**Apakah terdapat korelasi antara tingkat pendapatan dengan probabilitas melakukan gagal bayar pinjaman?**

In [96]:
# Periksa data tingkat pendapatan dan data gagal bayar pinjaman
income_debt=pd.pivot_table(credit,index='income_group', columns='debt', values='gender',aggfunc='count').fillna(0)
income_debt


debt                  0     1
income_group                 
lower income        868    58
middle income     13361  1247
middle up income   4254   344
rich                357    22
upper income        872    70

In [97]:
# Hitung persentase gagal bayar berdasarkan tingkat pendapatan
income_debt['total']=income_debt[0]+income_debt[1]
income_debt['percent 0']=round(income_debt[0]/income_debt['total']*100,1).astype(str)+'%'
income_debt['percent 1']=round(income_debt[1]/income_debt['total']*100,1).astype(str)+'%'
income_debt


debt                  0     1  total percent 0 percent 1
income_group                                            
lower income        868    58    926     93.7%      6.3%
middle income     13361  1247  14608     91.5%      8.5%
middle up income   4254   344   4598     92.5%      7.5%
rich                357    22    379     94.2%      5.8%
upper income        872    70    942     92.6%      7.4%

**Kesimpulan**

[Tulis kesimpulanmu berdasarkan manipulasi dan pengamatanmu.]
- Kelas middle memiliki persentase potensi gagal bayar hutang tertinggi '8.5%', dan kelas kaya memiliki persentase potensi gagal bayar hutang terendah '5.8%'. Sudah membentuk pola karena sesuai dengan logika

**Bagaimana tujuan kredit memengaruhi persentase gagal bayar?**

In [98]:
# Periksa persentase tingkat gagal bayar untuk setiap tujuan kredit dan lakukan penganalisisan
purpose_debt=pd.pivot_table(credit,index='purpose_group', columns='debt', values='gender',aggfunc='count').fillna(0)
purpose_debt


debt               0    1
purpose_group            
car             3903  403
education       3643  370
real estate    10028  782
wedding         2138  186

In [99]:
# Hitung persentase gagal bayar berdasarkan tujuan kredit
purpose_debt['total']=purpose_debt[0]+purpose_debt[1]
purpose_debt['percent 0']=round(purpose_debt[0]/purpose_debt['total']*100,1).astype(str)+'%'
purpose_debt['percent 1']=round(purpose_debt[1]/purpose_debt['total']*100,1).astype(str)+'%'
purpose_debt

debt               0    1  total percent 0 percent 1
purpose_group                                       
car             3903  403   4306     90.6%      9.4%
education       3643  370   4013     90.8%      9.2%
real estate    10028  782  10810     92.8%      7.2%
wedding         2138  186   2324     92.0%      8.0%

**Kesimpulan**

[Tulis kesimpulanmu berdasarkan manipulasi dan pengamatan yang kamu lakukan.]
- Potensi gagal bayar terbesar yakni untuk tujuan membeli mobil '9.4%', dan potensi terkecil untuk tujuan membeli real estate '7.2%'

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>

# Kesimpulan umum 

[Tuliskan kesimpulanmu di bagian akhir ini. Pastikan kamu memasukkan semua kesimpulan penting yang telah kamu buat berkaitan dengan caramu memproses dan menganalisis data. Kesimpulan tersebut harus membahas nilai yang hilang, duplikat, dan kemungkinan alasan, serta solusi untuk data bermasalah yang harus kamu tangani.]

[Tuliskan juga kesimpulanmu mengenai pertanyaan-pertanyaan yang diajukan di sini.]
- Pada kasus ini, dataset yang diberikan memiliki nilai yang hilang pada karakteristik 'days_employed' dan 'total_income' hal ini bisa terjadi karena asumsi saya calon nasabah menganggap kedua hal tersebut sebagai privasi mereka sehingga mengosongkan kolom tersebut, serta terdapat nilai yang terduplikasi pada karakteristik 'education' yang terjadi karena calon nasabah ada yang menulis tingkat pendidikan mereka menggunakan huruf kecil dan adapula yang menggunakan huruf kapital. Langkah yang dilakukan agar mendapatkan hasil analisa yang baik yakni memperbaiki nilai yang hilang dengan menggunakan metode median karena outlier cukup signifikan, dan untuk memperbaiki nilai yang terduplikasi dengan mengubah semua hurufnya menjadi huruf kecil.

- Beberapa cara telah dilakukan agar membentuk suatu pola untuk mengidentifikasi apakah calon nasabah tersebut layak diberikan kredit atau tidak. Hal ini dilakukan dengan cara mengkorelasikan beberapa karakteristik dengan history gagal bayar hutang mereka, mulai dari jumlah anak, status keluarga, tingkat pendapatan, dan tujuan kredit. Dari 

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, code yang dijalankan sudah sesuai dengan instruksi dari project.

</div>